### PR TITLE
[ENH] Make sure workflows do not run again for participant/sessions that already have expected output

### LINKED ISSUE #246: Make sure workflows don't run for participant/sessions that already have expected output
- `DicomReorgWorkflow` runs for all `downloaded` participant/session pairs **that have not been `organized`**
- `BidsConversionWorkflow` runs for all `organized` participant/session pairs **that have not been `bidsified`**
- `PipelineRunner` workflow runs for all `bidsified` participant/session pairs **that do not have `SUCCESS` in the derivatives `bagel.csv`**
  - This one is trickier since the bagel is not necessarily up to date...
- `PipelineTracker` workflow runs for all `bidsified` participant/session pairs

### PR DESCRIPTION
**Pull request by michellewang**
Closes #246

Changes:
- `DicomReorgWorkflow` runs for all **downloaded** participant/session pairs that **have *not* been organized**
- `BidsConversionWorkflow` runs for all **organized** participant/session pairs that **have *not* been bidsified**
- `PipelineRunner` workflow runs for all **bidsified** participant/session pairs that **do *not* have `SUCCESS`** in the imaging bagel
- `PipelineTracker` workflow runs for all **bidsified** participant/session pairs

TODOs
- [ ] Once #238 is merged we should probably add a call to `check_pipeline_version` in `PipelineRunner.get_participants_sessions_to_run()` and so that this doesn't fail if the user does not specify the version

### CODE REVIEW

**Comment by michellewang** [comment]
- [ ] Once #238 is merged we should probably add a call to `check_pipeline_version` in `PipelineRunner` and so that this doesn't fail if the user does not specify the version

File: nipoppy_cli/nipoppy/workflows/pipeline.py [hunk 1]
```diff
@@ -359,26 +359,23 @@ def run_cleanup(self, **kwargs):
...
-        # 2. "downloaded" but not "organized" (from doughnut)
-        # 3. "organized" but not "bidsified" (from doughnut)
-        # 4. "bidsified" but not completed (from doughnut/bagel)
-        # 5. "dataset" (i.e. apply on entire dataset, do not loop over anything)
-
-        # for now just check the participants/sessions that have BIDS data
-        return self.doughnut.get_bidsified_participants_sessions(
-            participant=participant, session=session
-        )
+        """
```
Comment (nikhil153) [inline nipoppy_cli/nipoppy/workflows/pipeline.py:366]: Do we need to specify `STEP` here for proper filtering of the bagel? Presumably this is applicable to all bids and proc pipelines?

File: nipoppy_cli/nipoppy/workflows/runner.py [hunk 2]
```diff
@@ -146,6 +147,35 @@ def launch_boutiques_run(
...
+    def get_participants_sessions_to_run(
+        self, participant: Optional[str], session: Optional[str]
+    ):
+        """Generate a list of participants and sessions to run.
+
+        Specifically, this list will include participants who have BIDS data but
+        who have not previously successfully completed the pipeline (according)
+        to the bagel file.
+        """
+        self.check_pipeline_version()  # in case this is called outside of run()
```
Comment (nikhil153) [inline nipoppy_cli/nipoppy/workflows/runner.py:159]: Do we need to check the `STEP` here as well?

File: nipoppy_cli/nipoppy/workflows/tracker.py [hunk 3]
```diff
@@ -58,14 +58,21 @@ def check_status(self, relative_paths: StrOrPathLike):
...
 
-            # TODO handle potentially zipped archives
             matches = list(self.dpath_pipeline_output.glob(str(relative_path)))
             self.logger.debug(f"Matches: {matches}")
             if not matches:
                 return Bagel.status_fail
 
         return Bagel.status_success
 
+    def get_participants_sessions_to_run(
```
Comment (nikhil153) [inline nipoppy_cli/nipoppy/workflows/tracker.py:68]: Same as about - do we need to include `STEP`?

File: nipoppy_cli/tests/test_workflow_pipeline.py [hunk 4]
```diff
@@ -47,8 +47,16 @@ def __init__(
             dry_run=dry_run,
         )
 
+    def get_participants_sessions_to_run(
```
Comment (nikhil153) [inline nipoppy_cli/tests/test_workflow_pipeline.py:50]: Do we need to a test for `STEP`?

**Review by nikhil153 — APPROVED** [review]
Only one important issue on handling of `STEPS`. If that's not relevant to this PR, then let's go ahead and merge!

File: nipoppy_cli/nipoppy/tabular/bagel.py [hunk 5]
```diff
@@ -104,3 +104,36 @@ class Bagel(BaseTabular):
...
+
+        bagel_subset = self.loc[
+            (self[self.col_pipeline_name] == pipeline_name)
+            & (self[self.col_pipeline_version] == pipeline_version)
+            & (self[self.col_participant_id].isin(participants))
+            & (self[self.col_session].isin(session))
+            & (self[self.col_pipeline_complete] == self.status_success)
+        ]
+
+        yield from bagel_subset[[self.col_participant_id, self.col_session]].itertuples(
```
[the remark this answers is not in the document]
Comment (michellewang) [reply nipoppy_cli/nipoppy/tabular/bagel.py:137]: Yes to save on memory -- current bagels are not huge but maybe they will grow!

**Comment by michellewang** [comment]
@nikhil153 I say let's not have `pipeline_step` in bagels just now -- would probably require a discussion with @alyssadai sometime but it's not a huge priority. Currently we can only track at the level of pipelines, not pipeline steps

**Pull request merged**